Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,560 @@
+```
+
+#### Concatenation
+
+Concatenating a sequence of patterns, _r0_, _r1_, _r2_, ..., will cause the
+resulting capture type to reflect the _concatenated capture type_, represented
+as a tuple of capture types or a single capture type depending on the overall
+quantity of captures in _r0_, _r1_, _r2_, ... If the overall capture quantity is
+`1`, the resulting capture type is the capture type of the single pattern that
+has a capture; otherwise, the resulting capture type is a tuple of capture types
+of all patterns that have a capture.
+
+```
+pattern ::= pattern pattern
+```
+
+```swift
+let graphemeBreakLowerBound = /([0-9a-fA-F]+)\.\.[0-9a-fA-F]+/
+// => `Regex<Substring>`
+
+// Result builder DSL equivalent:
+//     let graphemeBreakLowerBound = Regex {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//         ".."
+//         OneOrMore(CharacterClass.hexDigit)
+//     }
+//     // `.Captures == Substring`
+
+let graphemeBreakRange = /([0-9a-fA-F]+)\.\.([0-9a-fA-F]+)/
+// => `Regex<(Substring, Substring)>`
+
+// Result builder DSL equivalent:
+//     let graphemeBreakRange = Regex {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//         ".."
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == (Substring, Substring)`
+```
+
+#### Named Capturing Group
+
+A named capturing group in a pattern with multiple captures causes the resulting
+tuple to have a tuple element label at the corresponding capture type position.
+When the pattern has only one capture, there will be no tuple element label
+because there are no 1-element tuples.
+
+```
+pattern ::= '(' '?' '<' name '>' pattern ')'
+```
+
+```swift
+let graphemeBreakLowerBound = /(?<lower>[0-9A-F]+)\.\.[0-9A-F]+/
+// => `Regex<Substring>`
+
+let graphemeBreakRange = /(?<lower>[0-9A-F]+)\.\.(?<upper>[0-9A-F]+)/
+// => `Regex<(lower: Substring, upper: Substring)>`
+```
+
+#### Non-capturing Group
+
+A non-capturing group's capture type is the capture type of its underlying pattern.
+
+```
+pattern ::= '(' '?' ':' pattern ')'
+```
+
+```swift
+let graphemeBreakLowerBound = /([0-9A-F]+)(?:\.\.([0-9A-F]+))?/
+// => `Regex<Void>`
+
+// Result builder DSL equivalent:
+//     let graphemeBreakLowerBound = Regex {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//         Optionally {
+//             ".."
+//             OneOrMore(CharacterClass.hexDigit).capture()
+//         }
+//     }
+//     // `.Captures == (Substring, Substring?)`
+```
+
+#### Nested capturing group
+
+When capturing group is nested within another capturing group, they count as two
+distinct captures in the order their left parenthesis first appears in the
+regular expression literal. This is consistent with PCRE and allows us to use
+backreferences (e.g. `\2`) with linear indices.
+
+```swift
+let graphemeBreakPropertyData = /(([0-9A-F]+)(\.\.([0-9A-F]+)))\s*;\s(\w+).*/
+// Positions in capture tuple: 0 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    4 ^~~~~
```

Comment:
In PCRE, capture group indices start from 1, `\0` (or `$0` when substituting) refers to the whole match.
```swift
let graphemeBreakPropertyData = /(([0-9A-F]+)(\.\.([0-9A-F]+)))\s*;\s(\w+).*/
// Positions in capture tuple:  2 ^~~~~~~~~~~   4 ^~~~~~~~~~~      5 ^~~~~
//                                         3 ^~~~~~~~~~~~~~~~~
//                             1 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The result builder DSL equivalent is missing a `.capture()`:
```swift
//     let graphemeBreakPropertyData = Regex {
//         Regex {
//             OneOrMore(CharacterClass.hexDigit).capture() // (2)
//             Regex {
//                 ".."
//                 OneOrMore(CharacterClass.hexDigit).capture() // (4)
//             }.capture() // (3)
//         }.capture() // (1)
//         Repeat(CharacterClass.whitespace)
//         ";"
//         CharacterClass.whitespace
//         OneOrMore(CharacterClass.word).capture() // (5)
//         Repeat(CharacterClass.any)
//     }
```

Reply:
Ya, in what we've laid out here there will be an inconsistency between back reference numbering `\1` and tuple number `.0`. I don't think it's as intuitive for new users, but we could always include the whole match in the capture type and put it at position `0` to align with PCRE.

Reply:
@natecook1000 and I have been talking about this. In the current draft the match algorithm is implicitly conceptualized as:

```swift
extension String {
  func match<R: RegexProtocol>(_ regex: R) -> (match: Substring, captures: R.Captures)?
}
```

If we flattened the captures into the resulting tuple, then the index based access would align with the backreference numbering:

```swift
// longhand spelling
extension String {
  func match<R: RegexProtocol>(_ regex: R) -> (match: Substring)? where R.Captures == Void
  func match<R: RegexProtocol, C0>(_ regex: R) -> (match: Substring, C0)? where R.Captures == C0
  func match<R: RegexProtocol, C0, C1>(_ regex: R) -> (match: Substring, C0, C1)? where R.Captures == (C0, C1)
  ...
}

// variadic generics spelling
extension String {
  func match<R: RegexProtocol, T...>(_ regex: R) -> (match: Substring, T...)? where R.Captures == (T...)
}
```

The main avenue I see this foreclosing is returning a richer type than a tuple from `match` that we can hang additional operations off of.

Reply:
Thank you for the internal discussion. A flattened match+captures tuple is what I thought of too initially. As for "returning a richer type", I guess it could still be done. With ~~variadic generics~~ tuple splatting spelling
```swift
struct Match<Captures> {
  let captures: (Substring, Captures...)
  var match: Substring { captures.0 }
}
```
You would be able to access the full match by either using `.match` or `.captures.0`. Without tuple splatting, you could either declare types on the lines of
```swift
struct Match0 {
  let captures: Substring
  var match: Substring { captures }
}
struct Match1<C1> {
  let captures: (Substring, C1)
  var match: Substring { captures.0 }
}
struct Match2<C1, C2> {
  let captures: (Substring, C1, C2)
  var match: Substring { captures.0 }
}
...

// longhand spelling
extension String {
  func match<R: RegexProtocol>(_ regex: R) -> Match0? where R.Captures == Void
  func match<R: RegexProtocol, C1>(_ regex: R) -> Match1<C1>? where R.Captures == C1
  func match<R: RegexProtocol, C1, C2>(_ regex: R) -> Match2<C1, C2>? where R.Captures == (C1, C2)
  ...
}
```
Or, more concisely, declare a provisional `Match_` type having `(Substring, T...)` as its generic parameter:
```swift
// Captures == (Substring, T...)
struct Match_<Captures> {
  let captures: Captures
  var match: Substring { /* unsafely point to `captures.0` and bound it to `Substring` */ }
}

// longhand spelling
extension String {
  func match<R: RegexProtocol>(_ regex: R) -> Match_<Substring>? where R.Captures == Void
  func match<R: RegexProtocol, C1>(_ regex: R) -> Match_<(Substring, C1)>? where R.Captures == C1
  func match<R: RegexProtocol, C1, C2>(_ regex: R) -> Match_<(Substring, C1, C2)>? where R.Captures == (C1, C2)
  ...
}
```
I expect the `Match_` type to not have public initializers, so it's safe to assume that its generic parameter will always be a tuple (or directly `Substring`) having `Substring` as its first element.

Reply:
Ya, actually as I mention below with `Alternation` we can even add `@dynamicMemberLookup` to the `Match` type to allow referencing captures by index or name directly. The only thing we'd lose with a nominal type is tuple destructuring... which isn't nothing.

I don't think we want the "match" to be part of the captures generic type, because when we get around the result builder syntax it won't compose as well.

So maybe it could look something like:

```swift
@dynamicMemberLookup
struct Match<Captures...> {
  let _storage: (Substring, Captures...)
  let match: Substring
  let captures: (Captures...)

  subscript<Value>(dynamicMember keyPath: KeyPath<(Substring, Captures...), Value>) -> Value {
    _storage[keyPath: keyPath]
  }
}

extension String {
  func match<R: RegexProtocol>(_ regex: R) -> Match<R.Captures>?
}
```

This would allow you to index on the match type where `0` is the match and `1` is the first capture which aligns with PCRE syntax. It would also allow you to name based access. And you could read `captures` if you wanted to use tuple de-structuring. The only weird thing is that once you access `captures`, the indices on the tuple will have shifted down by one.

Reply:
With `@dynamicMemberLookup`, a captured group labeled `match` would conflict with the `Match` property named `match`. In that regard, `.captures` seems unavoidable.